### PR TITLE
Handle stacky code

### PR DESCRIPTION
**Pull request by kripken**
This handles code that doesn't fit in our AST, by introducing a local and a block where needed.

### CODE REVIEW

File: test/stacky.wasm.fromBinary [hunk 1]
```diff
@@ -0,0 +1,21 @@
+(module
+  (type $0 (func (param i32 i32) (result i32)))
+  (memory $0 256 256)
+  (export "add" (func $0))
+  (func $0 (type $0) (param $var$0 i32) (param $var$1 i32) (result i32)
+    (local $2 i32)
```
Comment (kripken) [inline test/stacky.wasm.fromBinary]: this is the new local that was added

File: src/wasm/wasm-binary.cpp [hunk 2]
```diff
@@ -1319,6 +1319,32 @@ Expression* WasmBinaryBuilder::popExpression() {
...
+  // add elements until we find a non-void
+  std::vector<Expression*> expressions;
+  expressions.push_back(ret);
+  while (1) {
+    auto* curr = popExpression();
+    expressions.push_back(curr);
+    if (curr->type != none) break;
+  }
+  auto* block = builder.makeBlock();
+  while (!expressions.empty()) {
```
Comment (dschuff) [inline src/wasm/wasm-binary.cpp]: Could you just do `std::copy_backward(expressions.begin(), expressions.end(), block->list.end())`

File: check.py [hunk 3]
```diff
@@ -175,6 +175,19 @@
...
+print '\n[ checking wasm-dis on provided binaries... ]\n'
+
+for t in tests:
+  if t.endswith('.wasm') and not t.startswith('spec'):
+    print '..', t
+    t = os.path.join(options.binaryen_test, t)
+    cmd = WASM_DIS + [t]
+    actual = run_command(cmd)
+
+    expected = open(t + '.fromBinary').read()
```
Comment (dschuff) [inline check.py]: maybe do something like
```
with open(...) as f:
 if actual != f.read():
   fail(...)
```
to close the handle?

**Comment by dschuff** [comment]
It might be worth considering adding new IR node like first/reverse-block to avoid having to introduce locals in the future. If stacky code becomes common among binaryen's producers, or if we want to optimize code produced by stack-aware producers, then we'd probably be pessimizing just by parsing their code.

**Comment by wllang** [comment]
A first/reverse-block does not solve the problem in general when the `pick` operator is added. Adding `pick` solves the problem in general, and then rather than using a local variable the value is left on the stack and picked as needed, and in the AST it can simply be expressed as an assignment to a constant block scoped variable so the solution is very similar to what was done here but without actually adding a local variable.

I assume the trouble binary 'stacky' code was the following which can be expressed nicely in the AST as follows:
```
(get_local $var$0)
(set_local $var$0 (i32.const 100))
(get_local $var$1)
(i32.add)
=>
(set $c1 (get_local $var$0))
(set_local $var$0 (i32.const 100))
(i32.add $c1 (get_local $var$1))
```

So when the decoder finds that void operator while it is building an expression, it flushes all the pending pieces of expressions into these local constants (rather than adding local variables and a block). When the code pops these it just references them by name, so `$c1` here. If these were being used multiple times then all but the last reference where if was popped would be `(pick $c1)`. You might want to go one step further and hide the `$c1` versus `(pick $c1)` distinction and make that an encoding detail to make the code easier to modify and write. 

It works nicely, please please give it a try, see the merit in this path and then some of the other compelling design changes might be more obvious.

**Comment by dschuff** [comment]
It's certainly true that `pick` with local constants (I guess those are essentially just SSA values) would handle the case of void expressions like the example you mentioned. However I would be hesitant to design Binaryen's IR around it without more consensus about whether and when it would be added to wasm. We still have to primarily target wasm as it exists today and in the near future.

To adopt any IR construct we'd want to have the design goals of ensuring that it can naturally express everything expressible in wasm, and also be useful for optimizations, so that we can avoid regressing code parsed from wasm that uses those constructs, and produce them whenever it would be a performance or size win.

It's not my current goal to use Binaryen to motivate changes in wasm, although I do think it could be a useful platform to do that. If someone were to come up with something that met the aforementioned criteria (which basically mean it would be useful for producers), and showed some useful improvement in performance/size of real programs (or the addition of some new capability) and additionally wasn't to onerous for consumers to implement, then that really would be a compelling and useful argument for adding it to wasm.

**Comment by kripken** [comment]
@dschuff 

Hmm, when I try to use `copy_backward` as suggested, I get [this](https://gist.github.com/kripken/3832ecdbe0f184c1ca63f22f56199602). I try to specialize that template for the iterator class we have, in `src/mixed_arena.h`,

````
namespace std {
template<class T>
struct iterator_traits< typename ArenaVector<T>::Iterator > {
  typedef T value_type;
};
} // std
````

but then it complains `template parameters not deducible in partial specialization` about `T`. But `T` can't be a final class yet, as the iterator is over a templated class..? Reading a bunch of stackoverflow posts about related errors and creating custom iterators, I am only getting more confused. Is there something simple I'm missing?

> Too bad we don't have a good way to express stacky code in text

Yeah, it would be nice if we got around to writing a new parser for the final wasm text format. But wabt works great meanwhile :)

> It might be worth considering adding new IR node like first/reverse-block to avoid having to introduce locals in the future.

Also agreed, we should consider this. Really depends on how beneficial/common stacky code is.

**Comment by dschuff** [comment]
actually can you do `typedef typename T value_type` instead?

**Comment by kripken** [comment]
No difference. It just doesn't like T not being defined, I guess.

**Comment by kripken** [comment]
So yeah, maybe better to look into full iteration as a followup. Will need someone that knows more C++ wizardry than me ;)

**Pull request merged**